File: bench/cpp/ops/BenchJointBilateralFilter.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchJointBilateralFilterConfig.hpp"

#include <cvcuda/OpJointBilateralFilter.hpp>
#include <cvcuda/OpReformat.hpp>

#include <nvbench/nvbench.cuh>

template<typename T>
inline void jointbilateralfilter(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape      = benchutils::GetShape<3, int3>(state.get_string("shape"));
    const benchutils::InputKind inputKind  = benchutils::GetInputKind(state.get_string("inputKind"));
    auto                        layout     = benchutils::GetStringParam(state, "layout", "NHWC");
    int                         diameter   = benchutils::GetIntParam<int>(state, "diameter");
    auto                        sigmaSpace = static_cast<float>(state.get_float64("sigmaSpace"));
    float                       sigmaColor = -1.f;

    using BT = typename nvcv::cuda::BaseType<T>;
    int ch   = nvcv::cuda::NumElements<T>;

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";
    if (layout != "NHWC" && layout != "NCHW" && layout != "NCHW_FAKE")
    {
        state.skip("JointBilateralFilter benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return;
    }
    if (isFakePlanar && inputKind == benchutils::InputKind::VarShape)
    {
        state.skip("Fake-planar (NCHW_FAKE) JointBilateralFilter benchmark is tensor-only");
        return;
    }

    NVCVBorderType borderType = benchutils::GetBorderType(state.get_string("border"));

    state.add_global_memory_reads(shape.x * shape.y * shape.z * sizeof(T));
    state.add_global_memory_writes(shape.x * shape.y * shape.z * sizeof(T));

    cvcuda::JointBilateralFilter op;

    // clang-format off

    if (isFakePlanar) // tensor-only: planar->interleaved->jointbilateralfilter->interleaved->planar
    {
        nvcv::Tensor src     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interSrc({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interDst({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::LcgValues<BT>());

        cvcuda::Reformat reformatOp;
        benchutils::warmup_and_exec(state, BENCH_JOINTBILATERALFILTER_WARMUP_ITERATIONS,
            [&op, &reformatOp, &src, &interSrc, &interDst, &dst, &diameter, &sigmaColor, &sigmaSpace, &borderType](cudaStream_t s) {
                reformatOp(s, src, interSrc);
                op(s, interSrc, interSrc, interDst, diameter, sigmaColor, sigmaSpace, borderType);
                reformatOp(s, interDst, dst);
            });
    }
    else if (inputKind == benchutils::InputKind::Tensor)
    {
        nvcv::Tensor src = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor color = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::LcgValues<BT>());
        benchutils::FillTensor<BT>(color, benchutils::LcgValues<BT>());

        benchutils::warmup_and_exec(state, BENCH_JOINTBILATERALFILTER_WARMUP_ITERATIONS,
            [&op, &src, &color, &dst, &diameter, &sigmaColor, &sigmaSpace, &borderType](cudaStream_t s) {
                op(s, src, color, dst, diameter, sigmaColor, sigmaSpace, borderType);
            });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        nvcv::ImageBatchVarShape src(shape.x);
        nvcv::ImageBatchVarShape color(shape.x);
        nvcv::ImageBatchVarShape dst(shape.x);

        if (isPlanar)
        {
            benchutils::FillPlanarImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0});
            color.pushBack(src.begin(), src.end());
            benchutils::FillPlanarImageBatchLike<T>(dst, src);
        }
        else
        {
            benchutils::FillImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0},
                                          benchutils::LcgValues<T>());
            // color can alias src (guidance = input is valid use case)
            color.pushBack(src.begin(), src.end());
            // Use FillImageBatchLike to ensure dst has same per-sample shapes as src
            benchutils::FillImageBatchLike<T>(dst, src, [](const long4_16a &) { return T{0}; });
        }

        nvcv::Tensor diameterTensor({{shape.x}, "N"}, nvcv::TYPE_S32);
        nvcv::Tensor sigmaSpaceTensor({{shape.x}, "N"}, nvcv::TYPE_F32);
        nvcv::Tensor sigmaColorTensor({{shape.x}, "N"}, nvcv::TYPE_F32);

        benchutils::FillTensor<int>(diameterTensor, [&diameter](const long4_16a &){ return diameter; });
        benchutils::FillTensor<float>(sigmaSpaceTensor, [&sigmaSpace](const long4_16a &){ return sigmaSpace; });
        benchutils::FillTensor<float>(sigmaColorTensor, [&sigmaColor](const long4_16a &){ return sigmaColor; });

        benchutils::warmup_and_exec(state, BENCH_JOINTBILATERALFILTER_WARMUP_ITERATIONS,
            [&op, &src, &color, &dst, &diameterTensor, &sigmaColorTensor, &sigmaSpaceTensor, &borderType](cudaStream_t s) {
                op(s, src, color, dst, diameterTensor, sigmaColorTensor, sigmaSpaceTensor, borderType);
            });
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(jointbilateralfilter, NVBENCH_TYPE_AXES(BENCH_JOINTBILATERALFILTER_TYPES))
BENCH_JOINTBILATERALFILTER_AXES;

File: bench/python/ops/bench_jointbilateralfilter.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA JointBilateralFilter operator benchmark - Python equivalent of BenchJointBilateralFilter.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_num_channels,
    get_dtype_size,
    get_format_from_dtype,
    get_border_type,
    create_tensor,
    create_image_batch_varshape,
    create_stream_cache,
    run_benchmark,
)


def jointbilateralfilter(state):
    """JointBilateralFilter operator benchmark matching C++ BenchJointBilateralFilter.cpp"""

    N, H, W = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    dtype = get_dtype(dtype_str)
    num_channels = get_num_channels(dtype_str)
    border = get_border_type(state.get_string("border"))
    input_kind = get_input_kind(state.get_string("inputKind"))
    try:
        layout = state.get_string("layout")
    except (KeyError, RuntimeError):
        layout = "NHWC"
    device_id = state.get_device()

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip(
            "JointBilateralFilter benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts"
        )
        return None
    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"
    if is_fake_planar and input_kind == "VarShape":
        state.skip(
            "Fake-planar (NCHW_FAKE) JointBilateralFilter benchmark is tensor-only"
        )
        return None

    diameter_scalar = int(state.get_int64("diameter"))
    sigma_space_scalar = float(state.get_float64("sigmaSpace"))
    sigma_color_scalar = -1.0

    dtype_size = get_dtype_size(dtype_str)
    state.add_global_memory_reads(N * H * W * dtype_size)
    state.add_global_memory_writes(N * H * W * dtype_size)

    get_stream = create_stream_cache()

    def do_jointbilateralfilter(
        dst, src, color, diameter, sigma_color, sigma_space, stream
    ):
        cvcuda.joint_bilateral_filter_into(
            dst,
            src,
            color,
            diameter,
            sigma_color,
            sigma_space,
            border=border,
            stream=stream,
        )

    if is_fake_planar:
        src = create_tensor(
            (N, num_channels, H, W),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode="lcg",
        )
        inter_src = create_tensor(
            (N, H, W, num_channels),
            dtype,
            device_id,
            layout="NHWC",
            fill_mode=0,
        )
        inter_dst = create_tensor(
            (N, H, W, num_channels), dtype, device_id, layout="NHWC", fill_mode=0
        )
        dst = create_tensor(
            (N, num_channels, H, W),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode=0,
        )

        def run_fake(launch):
            stream = get_stream(launch)
            cvcuda.reformat_into(inter_src, src, stream=stream)
            do_jointbilateralfilter(
                inter_dst,
                inter_src,
                inter_src,
                diameter_scalar,
                sigma_color_scalar,
                sigma_space_scalar,
                stream,
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    if input_kind == "Tensor":
        tensor_shape = (N, num_channels, H, W) if is_planar else (N, H, W, num_channels)
        src = create_tensor(
            tensor_shape,
            dtype,
            device_id,
            layout=layout,
            fill_mode="lcg",
        )
        color = create_tensor(
            tensor_shape,
            dtype,
            device_id,
            layout=layout,
            fill_mode="lcg",
        )
        dst = create_tensor(
            tensor_shape,
            dtype,
            device_id,
            layout=layout,
            fill_mode=0,
        )

        def run(launch):
            do_jointbilateralfilter(
                dst,
                src,
                color,
                diameter_scalar,
                sigma_color_scalar,
                sigma_space_scalar,
                get_stream(launch),
            )

        return run
    else:  # ImageBatchVarShape mode
        diameter = create_tensor(
            (N,), cvcuda.Type.S32, device_id, layout="N", fill_mode=diameter_scalar
        )
        sigma_space = create_tensor(
            (N,), cvcuda.Type.F32, device_id, layout="N", fill_mode=sigma_space_scalar
        )
        sigma_color = create_tensor(
            (N,), cvcuda.Type.F32, device_id, layout="N", fill_mode=sigma_color_scalar
        )

        img_format = get_format_from_dtype(dtype_str, num_channels, planar=is_planar)
        src = create_image_batch_varshape(
            (N, H, W, num_channels),
            0,
            img_format,
            dtype,
            device_id,
            fill_mode="lcg",
        )
        color = src
        dst = create_image_batch_varshape(
            (N, H, W, num_channels),
            0,
            img_format,
            dtype,
            device_id,
            fill_mode=0,
        )

        def run(launch):
            do_jointbilateralfilter(
                dst,
                src,
                color,
                diameter,
                sigma_color,
                sigma_space,
                get_stream(launch),
            )

        return run


if __name__ == "__main__":
    run_benchmark("jointbilateralfilter", jointbilateralfilter)
